server/src/instant/jdbc/sql.clj

```diff
@@ -152,11 +173,36 @@
                             :attributes (span-attrs ~'conn ~'query ~'tag)}
           (try
             (io/tag-io
-              (~query-fn ~'conn ~'query (merge ~opts
-                                               {:timeout *query-timeout-seconds*})))
+              (let [create-connection?# (not (instance? Connection ~'conn))
+                    opts# (merge ~opts
+                                 {:timeout *query-timeout-seconds*})
+                    c# (if create-connection?#
+                         (next-jdbc/get-connection ~'conn)
+                         ~'conn)]
+                (try
+                  (with-open [ps# (next-jdbc/prepare c# ~'query opts#)
+                              _cleanup# (register-in-progress c# ps#)]
+                    (~query-fn ps# nil opts#))
+                  (finally
+                    ;; Don't close the connection if a java.sql.Connection was
+                    ;; passed in, or we'll end transactions before they're done.
+                    (when create-connection?#
+                      (.close c#))))))
             (catch PSQLException e#
               (throw (ex/translate-and-throw-psql-exception! e#)))))))))
 
+(defn next-do-execute!
```

Comment:
Created an issue on the `next-jdbc` repo: https://github.com/seancorfield/next-jdbc/issues/287